core/src/main/java/com/linecorp/armeria/server/ParameterizedPathMapping.java

```diff
@@ -46,7 +46,14 @@
  */
 final class ParameterizedPathMapping extends AbstractPathMapping {
 
-    private static final Pattern VALID_PATTERN = Pattern.compile("(/[^/{}:]+|/:[^/{}]+|/\\{[^/{}]+})+/?");
+    private static final Pattern VALID_PATTERN = Pattern.compile(
+            '(' +
+            // If the segment doesn't start with ':' or '{', the behavior should be the same as ExactPathMapping
+            "/[^:{][^/]*|" +
+            "/:[^/{}]+|" +
```

Comment:
Question) 
1. Asumming that a path pattern is `/a/:b:c/:d`, `b:c` is used as a path variable, e.g. `ctx.pathParam("b:c")`. Is that an expected behavior?
2. `/brace/{b}:def` is an invalid pattern. `/colon/:a:verb` is an valid pattern and a path segment is captured with `a:verb` at which I was a bit suprised. 

If we don't support `/:param:verb` style in the `ParameterizedPathMapping`, I perfer to raise an exception when a path variable includes `:`.

Reply:
I also found that the handling of `{}` and `:` is different which confused me at first.
Having said this, I decided to just follow the previous regex pattern since I didn't want to introduce inadvertent breaking changes.

> Asumming that a path pattern is /a/:b:c/:d, b:c is used as a path variable, e.g. ctx.pathParam("b:c"). Is that an expected behavior?
> /brace/{b}:def is an invalid pattern. /colon/:a:verb is an valid pattern and a path segment is captured with a:verb at which I was a bit suprised.
> If we don't support /:param:verb style in the ParameterizedPathMapping, I prefer to raise an exception when a path variable includes :. What do you think?

I wanted to avoid introducing more breaking changes since this PR is already introducing certain breaking changes. I also agree that the assymmetry is confusing and it is easier to reason about if `:` and `{}` have the same treatment.

I guess this comes down to "does this issue annoy maintainers enough to make a breaking change (which I'm not sure if there is any impact)". I think we can discuss this further at the next stand up meeting.

Reply:
> Double checked that this behavior is consistent with the main branch.

I was confused. `/brace/{b}:def` only works on the main branch but didn't work with the previous release version, 1.28.4. As it is consistent with the previous version, it is okay.

This PR is good as is. However, ultimately, it would be good to support a path pattern in which the path variable can be located in a path segment. Some URL patterns seem useful:
- `/{project}/{repoName}.git`
- `/download/{filename}.pdf`

Reply:
>  However, ultimately, it would be good to support a path pattern that allows the path variable can be located in a path segment

I agree this could be useful. I think it's probably doable if we start using `{}` instead of `\0` to internally representing the parameterized variables. Additionally, we probably need to modify the route parsing logic at `Routers` as well.

I don't have immediate plans to handle this since `RegexPathMapping` can kind of handle this case - I'd be happy to review if you would like to work on it though.

Reply:
To support complex path patterns to build finer grain routes, I am considering dropping `:` from the path variable patterns in Armeria 2.0. We don't debate about that though. 

I don't see it was a good choice that we used `:` as a path variable. `:` seems simple but it can be also used as a literal. If a path variable is mixed with a literal `:`, it is ambiguous and difficult to know the exact name of the variable.

> I don't have immediate plans to handle this since RegexPathMapping can kind of handle this case - I'd be happy to review if you would like to work on it though.

I'll try it next time. 😃